source/extensions/filters/common/ratelimit/ratelimit_impl.cc

```diff
@@ -128,11 +128,10 @@ ClientPtr rateLimitClient(Server::Configuration::FactoryContext& context,
                           envoy::config::core::v3::ApiVersion transport_api_version) {
   // TODO(ramaraochavali): register client to singleton when GrpcClientImpl supports concurrent
   // requests.
-  const auto async_client_factory =
-      context.clusterManager().grpcAsyncClientManager().factoryForGrpcService(
-          grpc_service, context.scope(), true);
   return std::make_unique<Filters::Common::RateLimit::GrpcClientImpl>(
-      async_client_factory->create(), timeout, transport_api_version);
+      context.clusterManager().grpcAsyncClientManager().getOrCreateRawAsyncClient(
```

Comment:
I think this implies that multiple requests (filter-chains) on the same worker thread will use the same gRPC client to the rate limit service, instead of per-filter-chain gRPC client to the rate limit service. Will this change the rate limiting behavior?

Reply:
That's a good point. I haven't thought about it before. Will investigate.

Reply:
Any more thoughts on this? Should this be a configuration option?

@htuch for strategic opinions on this.

Reply:
I will add some tests to verify the ratelimit behavior first and see whether we need to make some change.

Reply:
I took a look at the implementation of RateLimit::GrpcClientImpl, sharing RawAsyncClient across GrpcClientImpl should be fine as long as GrpcClientImpl itself is not shared. I will add some test to verify it.

Reply:
To clarify, by "should be fine" do you mean "requests will flow" or do you mean that the rate-limiting for the different gRPC users will be independent?

Reply:
I guess it should be independent for each request.

Reply:
...should we have a test for that?

Reply:
Sure, I am working on it.

Reply:
I just pushed a test to verify the behavior of ratelimit filter, sharing raw async client across ratelimit filter doesn't affect the envoy grpc, but the test will be flaky for google grpc.(I suspect that it may have some race condition because google grpc has its own threading while envoy grpc doesn't). When the sharing of raw grpc async client is disabled, the test can pass reliably.

Reply:
What I was thinking for a test was that we would have 2 unrelated clients: one  sending frequent traffic, the other sporadic. The frequent-traffic client alone would be rate-limited. The sporadic-traffic alone would not be. I think the test should verify that when both those clients are sharing a cache, their rate-limiting behaviors would be the same as before.

Is that what you've tested?

Reply:
I initialize 4 http requests, each with their own GrpcClientImpl, they share the same raw async client to send their grpc ratelimit request, then I send ratelimit response to each ratelimit request to see if the behavior is expected(the upstream response is successful or failure).

Reply:
The flakiness is because of testing upstream setup, fixed with suggestion by Adi to use autonomous upstream.